File: src/commands/shell.ts
import { defineCommand } from 'citty'
import { consola } from 'consola'
import { composeFor } from './_compose.ts'
import { loadAppOrExit } from './_ctx.ts'

/**
 * `jib exec <app> [service] -- <cmd...>` and `jib run <app> <service> [-- <cmd...>]`.
 * citty's flag parser would eat the `--` separator, so we read from `rawArgs`
 * directly and do the parsing ourselves. Matches the Go implementation's
 * `parseExecArgs` / `parseRunArgs` semantics exactly.
 */

interface ExecParts {
  app: string
  service: string
  cmd: string[]
}

export function parseExecArgs(raw: string[]): ExecParts {
  if (raw.length === 0)
    throw new Error('missing app name — usage: jib exec <app> [service] -- <cmd>')
  const [app, ...rest] = raw
  const dash = rest.indexOf('--')
  if (dash === -1) {
    if (rest.length === 0)
      throw new Error('command required after app — usage: jib exec <app> [service] -- <cmd>')
    return { app: app as string, service: rest[0] as string, cmd: rest.slice(1) }
  }
  const before = rest.slice(0, dash)
  const after = rest.slice(dash + 1)
  return { app: app as string, service: (before[0] ?? '') as string, cmd: after }
}

export function parseRunArgs(raw: string[]): ExecParts {
  if (raw.length < 2)
    throw new Error('app name and service required — usage: jib run <app> <service> [-- <cmd>]')
  const [app, service, ...rest] = raw
  const dash = rest.indexOf('--')
  const cmd = dash === -1 ? rest : rest.slice(dash + 1)
  return { app: app as string, service: service as string, cmd }
}

async function handle(parts: ExecParts, mode: 'exec' | 'run'): Promise<void> {
  const { cfg, paths } = await loadAppOrExit(parts.app)
  const compose = composeFor(cfg, paths, parts.app)
  if (mode === 'exec') await compose.exec(parts.service, parts.cmd)
  else await compose.run(parts.service, parts.cmd)
}

export const execCmd = defineCommand({
  meta: { name: 'exec', description: 'Execute command in a running container' },
  async run({ rawArgs }) {
    if (rawArgs.includes('--help') || rawArgs.includes('-h')) {
      consola.log('usage: jib exec <app> [service] -- <cmd>')
      return
    }
    try {
      await handle(parseExecArgs(rawArgs), 'exec')
    } catch (err) {
      consola.error(err instanceof Error ? err.message : String(err))
      process.exit(1)
    }
  },
})

export const runCmd = defineCommand({
  meta: { name: 'run', description: 'Run a one-off command in a new container' },
  async run({ rawArgs }) {
    if (rawArgs.includes('--help') || rawArgs.includes('-h')) {
      consola.log('usage: jib run <app> <service> [-- <cmd>]')
      return
    }
    try {
      await handle(parseRunArgs(rawArgs), 'run')
    } catch (err) {
      consola.error(err instanceof Error ? err.message : String(err))
      process.exit(1)
    }
  },
})

File: _legacy/cmd/jib/shell.go
package main

import (
	"context"
	"fmt"

	"github.com/spf13/cobra"
)

// registerShellCommands wires up the per-app shell-access verbs: exec (into
// a running container) and run (ephemeral container for one-off tasks).
// Both use DisableFlagParsing so the raw argv reaches the handler — users
// pass a literal "--" to separate the target service from the command to
// run, which cobra's own flag parsing would otherwise eat.
func registerShellCommands(rootCmd *cobra.Command) {
	rootCmd.AddCommand(&cobra.Command{
		Use:                "exec <app> [service] -- <cmd>",
		Short:              "Execute command in running container",
		DisableFlagParsing: true,
		RunE:               runExec,
	})

	rootCmd.AddCommand(&cobra.Command{
		Use:                "run <app> <service> [-- <cmd>]",
		Short:              "Run a one-off command in a new container",
		DisableFlagParsing: true,
		RunE:               runRun,
	})
}

// parseExecArgs parses "jib exec <app> [service] -- <cmd...>" from raw args.
// Since DisableFlagParsing is true, we receive the full args slice.
func parseExecArgs(args []string) (appName, service string, cmdArgs []string, err error) {
	if len(args) == 0 {
		return "", "", nil, fmt.Errorf("missing app name\n\nUsage:\n  jib exec <app> [service] -- <cmd>")
	}

	appName = args[0]
	rest := args[1:]

	// Find "--" separator
	dashIdx := -1
	for i, a := range rest {
		if a == "--" {
			dashIdx = i
			break
		}
	}

	if dashIdx == -1 {
		// No "--", treat everything after app as command (service defaults to first)
		if len(rest) == 0 {
			return "", "", nil, fmt.Errorf("command is required after app name\n\nUsage:\n  jib exec <app> [service] -- <cmd>")
		}
		// If only one arg after app, it could be service name or command
		// Convention: if there's no --, first arg is service, rest is command
		service = rest[0]
		cmdArgs = rest[1:]
		return appName, service, cmdArgs, nil
	}

	// Has "--" separator
	beforeDash := rest[:dashIdx]
	afterDash := rest[dashIdx+1:]

	if len(beforeDash) > 0 {
		service = beforeDash[0]
	}
	cmdArgs = afterDash

	return appName, service, cmdArgs, nil
}

func runExec(cmd *cobra.Command, args []string) error {
	// DisableFlagParsing is true, so we must handle --help manually.
	for _, a := range args {
		if a == "--help" || a == "-h" {
			return cmd.Help()
		}
	}

	appName, service, cmdArgs, err := parseExecArgs(args)
	if err != nil {
		return err
	}

	cfg, err := loadConfig()
	if err != nil {
		return fmt.Errorf("loading config: %w", err)
	}

	compose, err := newCompose(cfg, appName)
	if err != nil {
		return err
	}

	return compose.Exec(context.Background(), service, cmdArgs)
}

// parseRunArgs parses "jib run <app> <service> [-- <cmd...>]" from raw args.
func parseRunArgs(args []string) (appName, service string, cmdArgs []string, err error) {
	if len(args) < 2 {
		return "", "", nil, fmt.Errorf("app name and service are required\n\nUsage:\n  jib run <app> <service> [-- <cmd>]")
	}

	appName = args[0]
	service = args[1]
	rest := args[2:]

	// Find "--" separator
	dashIdx := -1
	for i, a := range rest {
		if a == "--" {
			dashIdx = i
			break
		}
	}

	if dashIdx == -1 {
		cmdArgs = rest
	} else {
		cmdArgs = rest[dashIdx+1:]
	}

	return appName, service, cmdArgs, nil
}

func runRun(cmd *cobra.Command, args []string) error {
	// DisableFlagParsing is true, so we must handle --help manually.
	for _, a := range args {
		if a == "--help" || a == "-h" {
			return cmd.Help()
		}
	}

	appName, service, cmdArgs, err := parseRunArgs(args)
	if err != nil {
		return err
	}

	cfg, err := loadConfig()
	if err != nil {
		return fmt.Errorf("loading config: %w", err)
	}

	compose, err := newCompose(cfg, appName)
	if err != nil {
		return err
	}

	return compose.Run(context.Background(), service, cmdArgs)
}
